proteuslib/tools/parameter_sweep.py

```diff
@@ -36,48 +90,76 @@ def _init_mpi(mpi_comm=None):
+
+        for k, v in d.items():
+            # Build a vector of discrete values for this parameter
+            # and record the parameter's name
+            # v[1] = start, v[2] = stop, v[3] = resolution (number of elements)
+            # p = np.linspace(v[1], v[2], v[3])
+            p = v.sample(num_samples)
+            param_values.append(p)
+
+        num_var_params = len(param_values)
+
+        if sampling_type == "linear":
+            # Form an array with every possible combination of parameter values
+            global_combo_array = np.array(np.meshgrid(*param_values,indexing="ij"))
+            global_combo_array = global_combo_array.T.reshape(-1, num_var_params,order="F")
+        else:
+            sorting = np.argsort(param_values[0])
+            global_combo_array = np.column_stack(param_values)
+            global_combo_array = global_combo_array[sorting]
 
-    param_values = []
+    else:
+        global_combo_array = None
 
-    for k, v in d.items():
-        # Build a vector of discrete values for this parameter
-        # and record the parameter's name
-        # v[1] = start, v[2] = stop, v[3] = resolution (number of elements)
-        p = np.linspace(v[1], v[2], v[3])
-        param_values.append(p)
+    ### Broadcast the array to all processes
+    if num_procs > 1:
+        global_combo_array = comm.bcast(global_combo_array, root=0)
+    return global_combo_array
 
-    num_var_params = len(param_values)
+# ================================================================
 
-    # Form an array with every possible combination of parameter values
-    full_combo_array = np.array(np.meshgrid(*param_values))
-    full_combo_array = full_combo_array.T.reshape(-1, num_var_params)
+def _divide_combinations(global_combo_array, rank, num_procs):
 
     # Split the total list of combinations into NUM_PROCS chunks,
     # one per each of the MPI ranks
-    divided_combo_array = np.array_split(full_combo_array, num_procs, axis=0)
+    divided_combo_array = np.array_split(global_combo_array, num_procs, axis=0)
 
     # Return only this rank's portion of the total workload
     local_combo_array = divided_combo_array[rank]
 
-    return local_combo_array, full_combo_array
+    return local_combo_array
 
 # ================================================================
 
 def _update_model_values(m, param_dict, values):
 
     for k, item in enumerate(param_dict.values()):
 
-        param = item[0]
+        if isinstance(item,(tuple,list)):
+            param = item[0]
+        elif isinstance(item, Sample):
+            param = item.pyomo_object
+        else:
+            param = item
 
-        # Fix the single value to values[k]
         if param.is_variable_type():
+            # Fix the single value to values[k]
             param.fix(values[k])
 
         elif param.is_parameter_type():
             # Fix the single value to values[k]
             param.set_value(values[k])
 
+        elif param.is_indexed():
```

Comment:
Is there a use-case for this I'm missing? I took out explicit handling of indexed variables or parameters because it seemed like handing the parameter sweep tool an indexed pyomo object was likely to be an error. And if not, we need to handle this case in the `Sample` validation logic above.

Reply:
Yeah I had to add the indexed logic to work with the 0D RO model. For some reason RO.A_comp is an indexed Var() and wasn't being updated. How do you want to go about treating this?

Reply:
In this case, the user would unroll the index themselves. E.g., RO.A_comp[0] would work, I believe.

Reply:
FYI - A_comp is indexed by time and solvent component. I think this was originally done with dynamic modeling in mind (A_comp could change with time due to fouling although there are other ways we could account for membrane fouling over time). 

As for having the second index for solvent, this doesn't make much sense for RO since it supports only one solvent in the first place; I _think_ this was done for consistency with IDAES modeling and how things were normally indexed. On the other hand, for B_comp (salt permeability factor of membrane) this has indices for time and solute, but the RO model could technically take on more than one solute.

Reply:
This is exactly why I'm nervous about setting indexed things -- it could be the source of silent bugs (setting indices you didn't mean to set), which might be very hard to detect. Further, given the structure of IDAES models the only "proper" use case I could imagine is fixing all the time indices of an IndexedVar or IndexedParam to the same value. Fixing all the non-time indices, like for `flow_mass_phase_comp`, would almost surely be an error.

Reply:
I could be convinced that this could be a useful "typing aid" for the case of `A_comp`, e.g., and indexed var/param with exactly one index (in most use cases now). That said, we could check for that when we create the `Sample` object and replace the indexed component with the appropriate single var/param.

Reply:
I tried to take the indexed A_comp and get our the actual Var() I wanted to change, but I couldn't get it to work. Perhaps I was doing it wrong

Reply:
added the check for single valued indexed sets